### DISCUSSION TITLE
Royalities of a NFT collection

### DISCUSSION THREAD

**Discussion by Pelumi527** [aptos-labs/aptos-developer-discussions#117] [Questions] [answered]
Please How do I get the royalities fee of an NFT collection?

**Comment by 0x-j** [comment]
Hi, royalty can exist at both collection level and token level. We recommend you to always read it from the token level. Here's how you can get it with a view function. 

```rust
    #[view]
    public fun get_token_royalty(token_addr: address): (Option<u64>, Option<u64>, Option<address>) {
        let token_obj = object::address_to_object<token::Token>(token_addr);
        // royalty is optional, so you need to check if it exists before extract the value of it
        let maybe_royalty = royalty::get(token_obj);
        if (option::is_none(&maybe_royalty)) {
            (option::none(), option::none(), option::none())
        } else {
            let royalty = option::extract(&mut maybe_royalty);
            let numerator = royalty::numerator(&royalty);
            let denominator = royalty::denominator(&royalty);
            let payee_address = royalty::payee_address(&royalty);
            (option::some(numerator), option::some(denominator), option::some(payee_address))
        }
    }
```

[Source code](https://github.com/aptos-labs/aptos-core/blob/84433ec8ffc448f1a99478720a697c645ee08411/aptos-move/framework/aptos-token-objects/sources/royalty.move#L4) of royalty module if you are interested.

[Source code](https://github.com/0xaptosj/playground/blob/main/sources/nft_playground.move#L106-L119) for the view function code snippet. I also included a unit test calling it.

**Comment by gregnazario** [comment, marked as answer]
### How to get token royalties
There's an existing view function for this on the token module.  You can call this on a token directly.

https://github.com/aptos-labs/aptos-core/blob/84433ec8ffc448f1a99478720a697c645ee08411/aptos-move/framework/aptos-token-objects/sources/token.move#L464-L477

Source:
```move
    #[view]
    public fun royalty<T: key>(token: Object<T>): Option<Royalty> acquires Token {
        borrow(&token);
        let royalty = royalty::get(token);
        if (option::is_some(&royalty)) {
            royalty
        } else {
            let creator = creator(token);
            let collection_name = collection_name(token);
            let collection_address = collection::create_collection_address(&creator, &collection_name);
            let collection = object::address_to_object<collection::Collection>(collection_address);
            royalty::get(collection)
        }
    }
```

### How are royalties handled?

Royalties have two levels.  There's a token royalty, and a collection royalty.  The token one takes precedence over the collection one.

**Reply by Pelumi527** [reply]
Hi @gregnazario please I am trying to call this royality function on my frontend, but I am getting this error.
`{"error_code": "invalid_input", "message": "expected 1 type arguments for entry function 0x4::royalty::get, but got 0", "vm_error_code": null}`
Here is my payload 
```
const payload = {
      function: `0x4::royalty::get`,
      type_arguments: [
       
      ],
      arguments: [
        "0x311a3d5275b38e42a8b6ac3b92381a6887def12c446f474b2fdf0da04e8daa2c",
      ],
    };
```
